In [1]:
import pandas as pd
import numpy as np
import json

In [2]:
agency = pd.read_csv('gtfs_15_July/agency.txt')

calendar = pd.read_csv('gtfs_15_July/calendar.txt')

calendar_dates = pd.read_csv('gtfs_15_July/calendar_dates.txt')

routes = pd.read_csv('gtfs_15_July/routes.txt')

shapes = pd.read_csv('gtfs_15_July/shapes.txt')

stop_times = pd.read_csv('gtfs_15_July/stop_times.txt')

stops = pd.read_csv('gtfs_15_July/stops.txt')

trips = pd.read_csv('gtfs_15_July/trips.txt')

# Agency

In [3]:
agency

agency_id agency_name                          agency_url agency_timezone  \
0        978  Dublin Bus  https://www.transportforireland.ie   Europe/Dublin   

  agency_lang  
0          EN

# Calendar

In [4]:
calendar

service_id  monday  tuesday  wednesday  thursday  friday  saturday  sunday  \
0      y1003       1        1          1         1       1         0       0   
1      y1004       0        0          0         0       0         0       1   
2      y1005       0        0          0         0       0         1       0   
3      y1007       1        1          1         1       1         0       0   
4      y1008       1        0          0         0       0         0       1   
5      y1009       0        0          0         0       0         1       0   

   start_date  end_date  
0    20210715  20210724  
1    20210715  20210724  
2    20210715  20210724  
3    20210725  20211016  
4    20210725  20211016  
5    20210725  20211016

# Calendar Dates

In [5]:
calendar_dates

service_id      date  exception_type
0       y1007  20210802               2
1       y1008  20210726               2
2       y1008  20210809               2
3       y1008  20210816               2
4       y1008  20210823               2
5       y1008  20210830               2
6       y1008  20210906               2
7       y1008  20210913               2
8       y1008  20210920               2
9       y1008  20210927               2
10      y1008  20211004               2
11      y1008  20211011               2

# Routes

In [6]:
routes

route_id  agency_id route_short_name  route_long_name  route_type
0      60-1-b12-1        978                1              NaN           3
1    60-102-b12-1        978              102              NaN           3
2    60-104-b12-1        978              104              NaN           3
3     60-11-b12-1        978               11              NaN           3
4    60-111-b12-1        978              111              NaN           3
..            ...        ...              ...              ...         ...
281  60-757-d12-1        978              757              NaN           3
282   60-H1-d12-1        978               H1              NaN           3
283   60-H2-d12-1        978               H2              NaN           3
284   60-H3-d12-1        978               H3              NaN           3
285   60-H9-d12-1        978               H9              NaN           3

[286 rows x 5 columns]

In [7]:
routes.route_id.unique()

array(['60-1-b12-1', '60-102-b12-1', '60-104-b12-1', '60-11-b12-1',
       '60-111-b12-1', '60-114-b12-1', '60-116-b12-1', '60-118-b12-1',
       '60-120-b12-1', '60-122-b12-1', '60-123-b12-1', '60-13-b12-1',
       '60-130-b12-1', '60-14-b12-1', '60-140-b12-1', '60-142-b12-1',
       '60-145-b12-1', '60-14C-b12-1', '60-15-b12-1', '60-15D-b12-1',
       '60-150-b12-1', '60-155-b12-1', '60-151-b12-1', '60-15A-b12-1',
       '60-15B-b12-1', '60-16-b12-1', '60-161-b12-1', '60-16C-b12-1',
       '60-16D-b12-1', '60-17-b12-1', '60-17A-b12-1', '60-18-b12-1',
       '60-184-b12-1', '60-185-b12-1', '60-220-b12-1', '60-236-b12-1',
       '60-238-b12-1', '60-239-b12-1', '60-25-b12-1', '60-25A-b12-1',
       '60-25B-b12-1', '60-25D-b12-1', '60-25X-b12-1', '60-26-b12-1',
       '60-27-b12-1', '60-270-b12-1', '60-27A-b12-1', '60-27B-b12-1',
       '60-27X-b12-1', '60-29A-b12-1', '60-31-b12-1', '60-31A-b12-1',
       '60-31B-b12-1', '60-31D-b12-1', '60-32-b12-1', '60-32X-b12-1',
       '60-33-b12-1'

In [8]:
Current_routes = routes.route_short_name.unique()
Current_routes

array(['1', '102', '104', '11', '111', '114', '116', '118', '120', '122',
       '123', '13', '130', '14', '140', '142', '145', '14c', '15', '15d',
       '150', '155', '151', '15a', '15b', '16', '161', '16c', '16d', '17',
       '17a', '18', '184', '185', '220', '236', '238', '239', '25', '25a',
       '25b', '25d', '25x', '26', '27', '270', '27a', '27b', '27x', '29a',
       '31', '31a', '31b', '31d', '32', '32x', '33', '33a', '33b', '33e',
       '33x', '37', '38', '38a', '38b', '38d', '39', '39a', '39x', '4',
       '40', '40b', '40d', '40e', '41', '41a', '41b', '41c', '41d', '41x',
       '42', '42d', '43', '44', '44b', '45a', '46a', '46e', '47', '49',
       '51d', '51x', '53', '53a', '54a', '56a', '59', '6', '61', '63',
       '65', '65b', '66', '66a', '66b', '66e', '66x', '67', '67x', '68',
       '68x', '68a', '69', '69x', '7', '7a', '70', '70d', '747', '75',
       '76', '76a', '77a', '77x', '79', '79a', '7b', '7d', '8', '83',
       '83a', '84', '84a', '84x', '86', '9', '451

In [9]:
#converting all characters to uppercase to match the historical data 
x = map(lambda x:x.upper(),Current_routes)

In [10]:
Current_routes = list(x)

In [11]:
Current_routes

['1',
 '102',
 '104',
 '11',
 '111',
 '114',
 '116',
 '118',
 '120',
 '122',
 '123',
 '13',
 '130',
 '14',
 '140',
 '142',
 '145',
 '14C',
 '15',
 '15D',
 '150',
 '155',
 '151',
 '15A',
 '15B',
 '16',
 '161',
 '16C',
 '16D',
 '17',
 '17A',
 '18',
 '184',
 '185',
 '220',
 '236',
 '238',
 '239',
 '25',
 '25A',
 '25B',
 '25D',
 '25X',
 '26',
 '27',
 '270',
 '27A',
 '27B',
 '27X',
 '29A',
 '31',
 '31A',
 '31B',
 '31D',
 '32',
 '32X',
 '33',
 '33A',
 '33B',
 '33E',
 '33X',
 '37',
 '38',
 '38A',
 '38B',
 '38D',
 '39',
 '39A',
 '39X',
 '4',
 '40',
 '40B',
 '40D',
 '40E',
 '41',
 '41A',
 '41B',
 '41C',
 '41D',
 '41X',
 '42',
 '42D',
 '43',
 '44',
 '44B',
 '45A',
 '46A',
 '46E',
 '47',
 '49',
 '51D',
 '51X',
 '53',
 '53A',
 '54A',
 '56A',
 '59',
 '6',
 '61',
 '63',
 '65',
 '65B',
 '66',
 '66A',
 '66B',
 '66E',
 '66X',
 '67',
 '67X',
 '68',
 '68X',
 '68A',
 '69',
 '69X',
 '7',
 '7A',
 '70',
 '70D',
 '747',
 '75',
 '76',
 '76A',
 '77A',
 '77X',
 '79',
 '79A',
 '7B',
 '7D',
 '8',
 '83',
 '83A',
 '

In [12]:
len(Current_routes)

143

In [13]:
Trips_df = pd.read_csv("rt_trips_DB_2018.txt", ";")
Past_routes = Trips_df.LINEID.unique()
Past_routes

array(['68', '25B', '45A', '25A', '14', '77A', '39', '16', '40D', '27B',
       '142', '83', '130', '15', '46A', '33', '7', '39A', '49', '1',
       '123', '41', '67X', '59', '9', '40', '239', '76', '84', '53',
       '185', '151', '13', '15B', '65B', '29A', '61', '140', '79A', '38A',
       '31', '33B', '69', '44', '42', '67', '184', '238', '145', '17A',
       '32', '27A', '17', '27X', '18', '122', '54A', '66', '150', '56A',
       '37', '27', '15A', '65', '11', '47', '79', '83A', '63', '4', '120',
       '41C', '70', '84A', '220', '39X', '32X', '68A', '84X', '38', '102',
       '270', '51X', '33X', '75', '26', '66A', '31A', '111', '14C', '114',
       '76A', '44B', '161', '7A', '43', '25', '104', '33A', '16C', '42D',
       '31B', '66X', '31D', '33D', '41B', '40B', '7D', '46E', '38D',
       '118', '51D', '15D', '41A', '25D', '66B', '38B', '236', '7B',
       '41X', '69X', '68X', '25X', '40E', '70D', '116', '77X', '16D',
       '33E', '41D'], dtype=object)

In [14]:
len(Past_routes)

130

In [15]:
#Finding the new lines that have been added since 2018
difference = list(set(Current_routes) - set(Past_routes))
len(difference)

13

In [17]:
#Finding the new lines that have been added since 2018
difference2 = list(set(Past_routes) - set(Current_routes))
len(difference2)

0

In [16]:
difference

['53A',
 'H3',
 'H1',
 '757',
 '86',
 '747',
 '155',
 'H2',
 '8',
 '6',
 'H9',
 '451',
 '66E']

A total of 13 new lines have been added since 2018

In [18]:
routes.nunique()

route_id            286
agency_id             1
route_short_name    143
route_long_name       0
route_type            1
dtype: int64

In [19]:
routes['route_short_name'].value_counts()

33b    3
151    2
27x    2
13     2
27     2
      ..
47     2
270    2
14c    2
68x    2
33d    1
Name: route_short_name, Length: 143, dtype: int64

In [20]:
Trips_df.nunique()

DATASOURCE              1
DAYOFSERVICE          360
TRIPID             658964
LINEID                130
ROUTEID               588
DIRECTION               2
PLANNEDTIME_ARR     64461
PLANNEDTIME_DEP       791
ACTUALTIME_ARR      68122
ACTUALTIME_DEP      66771
BASIN                   1
TENDERLOT               0
SUPPRESSED              1
JUSTIFICATIONID      3526
LASTUPDATE            360
NOTE                46690
dtype: int64

# Shapes

In [21]:
shapes

shape_id  shape_pt_lat  shape_pt_lon  shape_pt_sequence  \
0       60-116-b12-1.56.O     53.329631     -6.249012                  1   
1       60-116-b12-1.56.O     53.329316     -6.248409                  2   
2       60-116-b12-1.56.O     53.329056     -6.247819                  3   
3       60-116-b12-1.56.O     53.328519     -6.246715                  4   
4       60-116-b12-1.56.O     53.328406     -6.246390                  5   
...                   ...           ...           ...                ...   
251075   60-H9-d12-1.96.I     53.350979     -6.252764                 87   
251076   60-H9-d12-1.96.I     53.350979     -6.252764                 88   
251077   60-H9-d12-1.96.I     53.350468     -6.255969                 89   
251078   60-H9-d12-1.96.I     53.350468     -6.255969                 90   
251079   60-H9-d12-1.96.I     53.348854     -6.256801                 91   

        shape_dist_traveled  
0                      0.00  
1                     53.15  
2                    101.85  
3                    196.45  
4                    221.44  
...                     ...  
251075              5277.76  
251076              5277.76  
251077              5497.96  
251078              5497.96  
251079              5685.81  

[251080 rows x 5 columns]

In [22]:
shapes.nunique()

shape_id                 719
shape_pt_lat           19816
shape_pt_lon           19816
shape_pt_sequence        720
shape_dist_traveled    73808
dtype: int64

# Stop Times


In [23]:
stop_times

trip_id arrival_time departure_time  \
0          7712.y1009.60-1-d12-1.1.O     18:40:00       18:40:00   
1          7712.y1009.60-1-d12-1.1.O     18:40:44       18:40:44   
2          7712.y1009.60-1-d12-1.1.O     18:41:21       18:41:21   
3          7712.y1009.60-1-d12-1.1.O     18:42:20       18:42:20   
4          7712.y1009.60-1-d12-1.1.O     18:42:58       18:42:58   
...                              ...          ...            ...   
1874914  6667.y1003.60-H9-b12-1.97.I     08:03:12       08:03:12   
1874915  6667.y1003.60-H9-b12-1.97.I     08:03:50       08:03:50   
1874916  6667.y1003.60-H9-b12-1.97.I     08:05:22       08:05:22   
1874917  6667.y1003.60-H9-b12-1.97.I     08:06:35       08:06:35   
1874918  6667.y1003.60-H9-b12-1.97.I     08:08:47       08:08:47   

              stop_id  stop_sequence  stop_headsign  pickup_type  \
0        8240DB000226              1     Sandymount            0   
1        8220DB000228              2     Sandymount            0   
2        8240DB000229              3     Sandymount            0   
3        8240DB000227              4     Sandymount            0   
4        8240DB000230              5     Sandymount            0   
...               ...            ...            ...          ...   
1874914  8220DB000619             16   Abbey Street            0   
1874915  8220DB000675             17   Abbey Street            0   
1874916  8220DB000620             18   Abbey Street            0   
1874917  8220DB007569             19   Abbey Street            0   
1874918  8220DB007591             20   Abbey Street            0   

         drop_off_type  shape_dist_traveled  
0                    0                 0.00  
1                    0               267.48  
2                    0               483.53  
3                    0               834.47  
4                    0              1063.49  
...                ...                  ...  
1874914              0              4745.60  
1874915              0              4900.43  
1874916              0              5277.76  
1874917              0              5497.96  
1874918              0              5685.81  

[1874919 rows x 9 columns]

In [24]:
stop_times.nunique()

trip_id                32683
arrival_time           77087
departure_time         77087
stop_id                 4208
stop_sequence            109
stop_headsign            132
pickup_type                2
drop_off_type              2
shape_dist_traveled    12430
dtype: int64

In [25]:
sorted_df = stop_times.groupby(["trip_id"]).apply(lambda x: x.sort_values(["stop_sequence"], ascending = True)).reset_index(drop=True)

In [27]:
sorted_df

trip_id arrival_time departure_time       stop_id  \
0         1.y1003.60-38-b12-1.178.I     08:06:00       08:06:00  8240DB004745   
1         1.y1003.60-38-b12-1.178.I     08:06:22       08:06:22  8240DB004989   
2         1.y1003.60-38-b12-1.178.I     08:08:04       08:08:04  8240DB004990   
3         1.y1003.60-38-b12-1.178.I     08:08:23       08:08:23  8240DB004991   
4         1.y1003.60-38-b12-1.178.I     08:08:39       08:08:39  8240DB004992   
...                             ...          ...            ...           ...   
1874914  9999.y1009.60-9-d12-1.12.O     20:57:21       20:57:21  8230DB002455   
1874915  9999.y1009.60-9-d12-1.12.O     20:57:52       20:57:52  8230DB002456   
1874916  9999.y1009.60-9-d12-1.12.O     20:58:21       20:58:21  8230DB002445   
1874917  9999.y1009.60-9-d12-1.12.O     20:58:48       20:58:48  8230DB002457   
1874918  9999.y1009.60-9-d12-1.12.O     20:59:19       20:59:19  8230DB004392   

         stop_sequence     stop_headsign  pickup_type  drop_off_type  \
0                    1   Burlington Road            0              0   
1                    2   Burlington Road            0              0   
2                    3   Burlington Road            0              0   
3                    4   Burlington Road            0              0   
4                    5   Burlington Road            0              0   
...                ...               ...          ...            ...   
1874914             62   Limekiln Avenue            0              0   
1874915             63   Limekiln Avenue            0              0   
1874916             64   Limekiln Avenue            0              0   
1874917             65   Limekiln Avenue            0              0   
1874918             66   Limekiln Avenue            0              0   

         shape_dist_traveled  
0                       0.00  
1                     238.57  
2                    1331.14  
3                    1537.99  
4                    1707.11  
...                      ...  
1874914             19651.68  
1874915             19897.53  
1874916             20128.70  
1874917             20343.74  
1874918             20592.16  

[1874919 rows x 9 columns]

# Trips

In [28]:
trips

route_id service_id                       trip_id  \
0      60-116-b12-1      y1003  3736.y1003.60-116-b12-1.57.I   
1      60-116-b12-1      y1003  3737.y1003.60-116-b12-1.56.O   
2      60-116-d12-1      y1007  3736.y1007.60-116-d12-1.56.I   
3      60-116-d12-1      y1007  3737.y1007.60-116-d12-1.55.O   
4      60-118-b12-1      y1003  3770.y1003.60-118-b12-1.58.I   
...             ...        ...                           ...   
32678   60-H9-b12-1      y1003   6666.y1003.60-H9-b12-1.97.I   
32679   60-H9-b12-1      y1003   6667.y1003.60-H9-b12-1.97.I   
32680   60-H9-d12-1      y1007   6663.y1007.60-H9-d12-1.96.I   
32681   60-H9-d12-1      y1007   6664.y1007.60-H9-d12-1.96.I   
32682   60-H9-d12-1      y1007   6665.y1007.60-H9-d12-1.96.I   

                shape_id                                      trip_headsign  \
0      60-116-b12-1.57.I  Church of Our Lady of Good Council - Parnell S...   
1      60-116-b12-1.56.O   Sussex Road - Church of Our Lady of Good Council   
2      60-116-d12-1.56.I  Church of Our Lady of Good Council - Parnell S...   
3      60-116-d12-1.55.O   Sussex Road - Church of Our Lady of Good Council   
4      60-118-b12-1.58.I        Enniskerry Road (Ballycorus Rd) - Eden Quay   
...                  ...                                                ...   
32678   60-H9-b12-1.97.I   All Saints Road - National Lottery Head Quarters   
32679   60-H9-b12-1.97.I   All Saints Road - National Lottery Head Quarters   
32680   60-H9-d12-1.96.I   All Saints Road - National Lottery Head Quarters   
32681   60-H9-d12-1.96.I   All Saints Road - National Lottery Head Quarters   
32682   60-H9-d12-1.96.I   All Saints Road - National Lottery Head Quarters   

       direction_id  
0                 1  
1                 0  
2                 1  
3                 0  
4                 1  
...             ...  
32678             1  
32679             1  
32680             1  
32681             1  
32682             1  

[32683 rows x 6 columns]

In [29]:
specific_service = trips.loc[trips['service_id']== 'y1003']

In [30]:
specific_service

route_id service_id                       trip_id  \
0      60-116-b12-1      y1003  3736.y1003.60-116-b12-1.57.I   
1      60-116-b12-1      y1003  3737.y1003.60-116-b12-1.56.O   
4      60-118-b12-1      y1003  3770.y1003.60-118-b12-1.58.I   
63      60-11-b12-1      y1003   1916.y1003.60-11-b12-1.18.I   
64      60-11-b12-1      y1003   1917.y1003.60-11-b12-1.17.O   
...             ...        ...                           ...   
32461   60-H3-b12-1      y1003  6742.y1003.60-H3-b12-1.354.I   
32462   60-H3-b12-1      y1003  6743.y1003.60-H3-b12-1.353.O   
32677   60-H9-b12-1      y1003   6665.y1003.60-H9-b12-1.97.I   
32678   60-H9-b12-1      y1003   6666.y1003.60-H9-b12-1.97.I   
32679   60-H9-b12-1      y1003   6667.y1003.60-H9-b12-1.97.I   

                shape_id                                      trip_headsign  \
0      60-116-b12-1.57.I  Church of Our Lady of Good Council - Parnell S...   
1      60-116-b12-1.56.O   Sussex Road - Church of Our Lady of Good Council   
4      60-118-b12-1.58.I        Enniskerry Road (Ballycorus Rd) - Eden Quay   
63      60-11-b12-1.18.I              Blackthorn Road - Saint Pappin's Road   
64      60-11-b12-1.17.O  Saint Pappin's Road - Blackthorn Drive (Blackt...   
...                  ...                                                ...   
32461  60-H3-b12-1.354.I      Howth Summit - National Lottery Head Quarters   
32462  60-H3-b12-1.353.O      National Lottery Head Quarters - Howth Summit   
32677   60-H9-b12-1.97.I   All Saints Road - National Lottery Head Quarters   
32678   60-H9-b12-1.97.I   All Saints Road - National Lottery Head Quarters   
32679   60-H9-b12-1.97.I   All Saints Road - National Lottery Head Quarters   

       direction_id  
0                 1  
1                 0  
4                 1  
63                1  
64                0  
...             ...  
32461             1  
32462             0  
32677             1  
32678             1  
32679             1  

[6743 rows x 6 columns]

In [31]:
specific_service.nunique()

route_id          102
service_id          1
trip_id          6743
shape_id          336
trip_headsign     177
direction_id        2
dtype: int64

In [32]:
trip_id_df = trips.loc[trips['trip_id']== '1.y1003.60-38-b12-1.178.I']

In [33]:
trip_id_df

route_id service_id                    trip_id           shape_id  \
15903  60-38-b12-1      y1003  1.y1003.60-38-b12-1.178.I  60-38-b12-1.178.I   

                                         trip_headsign  direction_id  
15903  Damastown Drive - Burlington Road (Mespil Road)             1

In [34]:
print(trips.nunique())

route_id           204
service_id           6
trip_id          32683
shape_id           719
trip_headsign      177
direction_id         2
dtype: int64


In [35]:
Trips_df = pd.read_csv("rt_trips_DB_2018.txt", ";")
Trips_df

DATASOURCE        DAYOFSERVICE   TRIPID LINEID  ROUTEID  DIRECTION  \
0               DB  07-FEB-18 00:00:00  6253783     68    68_80          1   
1               DB  07-FEB-18 00:00:00  6262138    25B  25B_271          2   
2               DB  07-FEB-18 00:00:00  6254942    45A   45A_70          2   
3               DB  07-FEB-18 00:00:00  6259460    25A  25A_273          1   
4               DB  07-FEB-18 00:00:00  6253175     14    14_15          1   
...            ...                 ...      ...    ...      ...        ...   
2182632         DB  14-MAY-18 00:00:00  6765849    123   123_36          2   
2182633         DB  14-MAY-18 00:00:00  6765469     75    75_17          1   
2182634         DB  14-MAY-18 00:00:00  6765486    33D   33D_62          2   
2182635         DB  14-MAY-18 00:00:00  6764987     70    70_60          1   
2182636         DB  14-MAY-18 00:00:00  6765012     27    27_19          1   

         PLANNEDTIME_ARR  PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP  \
0                  87245            84600         87524.0         84600.0   
1                  30517            26460         32752.0             NaN   
2                  35512            32100         36329.0         32082.0   
3                  57261            54420         58463.0         54443.0   
4                  85383            81600         84682.0         81608.0   
...                  ...              ...             ...             ...   
2182632            61560            57840         61365.0         57859.0   
2182633            53416            48600             NaN         48823.0   
2182634            29460            26400         29904.0             NaN   
2182635            65277            60600         66341.0             NaN   
2182636            47722            41700         47508.0         41642.0   

          BASIN  TENDERLOT  SUPPRESSED  JUSTIFICATIONID          LASTUPDATE  \
0        BasDef        NaN         NaN              NaN  28-FEB-18 12:05:11   
1        BasDef        NaN         NaN              NaN  28-FEB-18 12:05:11   
2        BasDef        NaN         NaN              NaN  28-FEB-18 12:05:11   
3        BasDef        NaN         NaN              NaN  28-FEB-18 12:05:11   
4        BasDef        NaN         NaN              NaN  28-FEB-18 12:05:11   
...         ...        ...         ...              ...                 ...   
2182632  BasDef        NaN         NaN              NaN  26-JUN-18 09:13:13   
2182633  BasDef        NaN         NaN              NaN  26-JUN-18 09:13:13   
2182634  BasDef        NaN         NaN              NaN  26-JUN-18 09:13:13   
2182635  BasDef        NaN         NaN              NaN  26-JUN-18 09:13:13   
2182636  BasDef        NaN         NaN              NaN  26-JUN-18 09:13:13   

              NOTE  
0        ,2967409,  
1        ,2580260,  
2        ,2448968,  
3        ,3094242,  
4        ,2526331,  
...            ...  
2182632  ,3216350,  
2182633  ,2865284,  
2182634  ,3077688,  
2182635  ,3208841,  
2182636  ,2960092,  

[2182637 rows x 16 columns]

In [36]:
Trips_df['LINEID'].value_counts()

46A    76728
145    62730
15     58083
40     56974
39A    55301
       ...  
68X      243
51X      238
16D       84
41D       36
33E       14
Name: LINEID, Length: 130, dtype: int64

In [37]:
Trips_df['ROUTEID'].value_counts()

46A_74     37182
46A_67     35759
145_102    29259
15_17      28471
145_105    27964
           ...  
140_16         5
41B_58         5
14C_10         4
41X_131        4
33D_76         2
Name: ROUTEID, Length: 588, dtype: int64

# Stops

In [3]:
stops

stop_id                      stop_name   stop_lat  stop_lon
0     8220DB000002    Parnell Square West, stop 2  53.352244 -6.263723
1     8220DB000003    Parnell Square West, stop 3  53.352309 -6.263811
2     8220DB000004    Parnell Square West, stop 4  53.352575 -6.264175
3     8220DB000006    Parnell Square West, stop 6  53.352749 -6.264454
4     8220DB000007    Parnell Square West, stop 7  53.352841 -6.264570
...            ...                            ...        ...       ...
4203  8350DB007461         Charlesland, stop 7461  53.128932 -6.062803
4204  8350DB007462         Charlesland, stop 7462  53.128801 -6.062480
4205  8350DB007574      Southern Cross, stop 7574  53.182348 -6.130064
4206  8350DB007823  Enniskerry Village, stop 7823  53.194198 -6.170184
4207  8530DB007824  Enniskerry Village, stop 7824  53.192558 -6.171137

[4208 rows x 4 columns]

In [4]:
stops.to_json()

'{"stop_id":{"0":"8220DB000002","1":"8220DB000003","2":"8220DB000004","3":"8220DB000006","4":"8220DB000007","5":"8220DB000008","6":"8220DB000010","7":"8220DB000011","8":"8220DB000012","9":"8220DB000014","10":"8220DB000015","11":"8220DB000016","12":"8220DB000017","13":"8220DB000018","14":"8220DB000019","15":"8220DB000021","16":"8220DB000022","17":"8220DB000023","18":"8220DB000024","19":"8220DB000025","20":"8220DB000027","21":"8220DB000028","22":"8220DB000029","23":"8220DB000030","24":"8220DB000031","25":"8220DB000032","26":"8220DB000033","27":"8220DB000035","28":"8220DB000036","29":"8220DB000037","30":"8220DB000038","31":"8220DB000039","32":"8220DB000040","33":"8220DB000041","34":"8220DB000042","35":"8220DB000043","36":"8220DB000044","37":"8220DB000045","38":"8220DB000046","39":"8220DB000047","40":"8220DB000048","41":"8220DB000049","42":"8220DB000051","43":"8220DB000052","44":"8220DB000053","45":"8220DB000054","46":"8220DB000055","47":"8220DB000056","48":"8220DB000057","49":"8220DB00005

# Merging of data

Merge stops with stop times on stopid to get the sequence of stops and their lat and long?

In [162]:
full_merge = pd.merge(stop_times, stops,on = 'stop_id') 

In [163]:
full_merge = pd.merge(full_merge, trips,on = 'trip_id')

In [164]:
full_merge = pd.merge(full_merge, routes,on = 'route_id')

In [165]:
pd.set_option('display.max_columns', None)
full_merge

trip_id arrival_time departure_time  \
0            7712.y1009.60-1-d12-1.1.O     18:40:00       18:40:00   
1            7712.y1009.60-1-d12-1.1.O     18:40:44       18:40:44   
2            7712.y1009.60-1-d12-1.1.O     18:41:21       18:41:21   
3            7712.y1009.60-1-d12-1.1.O     18:42:20       18:42:20   
4            7712.y1009.60-1-d12-1.1.O     18:42:58       18:42:58   
...                                ...          ...            ...   
1874914  1468.y1003.60-84A-b12-1.342.I     10:47:23       10:47:23   
1874915  1468.y1003.60-84A-b12-1.342.I     10:47:56       10:47:56   
1874916  1468.y1003.60-84A-b12-1.342.I     10:48:31       10:48:31   
1874917  1468.y1003.60-84A-b12-1.342.I     10:49:13       10:49:13   
1874918  1468.y1003.60-84A-b12-1.342.I     10:49:41       10:49:41   

              stop_id  stop_sequence      stop_headsign  pickup_type  \
0        8240DB000226              1         Sandymount            0   
1        8220DB000228              2         Sandymount            0   
2        8240DB000229              3         Sandymount            0   
3        8240DB000227              4         Sandymount            0   
4        8240DB000230              5         Sandymount            0   
...               ...            ...                ...          ...   
1874914  8250DB003155             29   St Vincents Hosp            0   
1874915  8250DB003156             30   St Vincents Hosp            0   
1874916  8250DB003158             31   St Vincents Hosp            0   
1874917  8250DB003159             32   St Vincents Hosp            0   
1874918  8250DB003160             33   St Vincents Hosp            0   

         drop_off_type  shape_dist_traveled                        stop_name  \
0                    0                 0.00         Shanard Avenue, stop 226   
1                    0               267.48          Shanliss Road, stop 228   
2                    0               483.53           Oldtown Road, stop 229   
3                    0               834.47         Shanliss Drive, stop 227   
4                    0              1063.49          Shanowen Road, stop 230   
...                ...                  ...                              ...   
1874914              0             11958.10       Beech Park Road, stop 3155   
1874915              0             12221.99            Deansworth, stop 3156   
1874916              0             12501.33      Deansgrange Road, stop 3158   
1874917              0             12909.46  Deansgrange Cemetery, stop 3159   
1874918              0             13128.38    St Fintan's Villas, stop 3160   

          stop_lat  stop_lon      route_id service_id            shape_id  \
0        53.391141 -6.262200    60-1-d12-1      y1009      60-1-d12-1.1.O   
1        53.391877 -6.259720    60-1-d12-1      y1009      60-1-d12-1.1.O   
2        53.391400 -6.256536    60-1-d12-1      y1009      60-1-d12-1.1.O   
3        53.391144 -6.251345    60-1-d12-1      y1009      60-1-d12-1.1.O   
4        53.389888 -6.249066    60-1-d12-1      y1009      60-1-d12-1.1.O   
...            ...       ...           ...        ...                 ...   
1874914  53.274256 -6.161937  60-84A-b12-1      y1003  60-84A-b12-1.342.I   
1874915  53.276538 -6.163024  60-84A-b12-1      y1003  60-84A-b12-1.342.I   
1874916  53.278938 -6.164181  60-84A-b12-1      y1003  60-84A-b12-1.342.I   
1874917  53.282461 -6.165920  60-84A-b12-1      y1003  60-84A-b12-1.342.I   
1874918  53.284379 -6.166752  60-84A-b12-1      y1003  60-84A-b12-1.342.I   

                                             trip_headsign  direction_id  \
0        Shanard Road (Shanard Avenue) - Saint John's C...             0   
1        Shanard Road (Shanard Avenue) - Saint John's C...             0   
2        Shanard Road (Shanard Avenue) - Saint John's C...             0   
3        Shanard Road (Shanard Avenue) - Saint John's C...             0   
4        Shanard Road (Shanard Avenue) - Saint John's C... 

In [166]:
full_merge_short = full_merge[['trip_id', 'stop_id','stop_sequence',
                               'stop_headsign','stop_name','stop_lat','stop_lon','direction_id','route_short_name']]

In [167]:
full_merge_short.nunique()

trip_id             32683
stop_id              4208
stop_sequence         109
stop_headsign         132
stop_name            4208
stop_lat             4208
stop_lon             4208
direction_id            2
route_short_name      103
dtype: int64

In [168]:
full_merge_short['stop_id']

0          8240DB000226
1          8220DB000228
2          8240DB000229
3          8240DB000227
4          8240DB000230
               ...     
1874914    8250DB003155
1874915    8250DB003156
1874916    8250DB003158
1874917    8250DB003159
1874918    8250DB003160
Name: stop_id, Length: 1874919, dtype: object

In [169]:
def short_id(df):
    stops = df['stop_id']
    short_stop = int(stops[len(stops) - 5:])
    return short_stop

In [170]:
pd.options.mode.chained_assignment = None  # default='warn'

full_merge_short['short_stop_id'] = full_merge_short.apply(short_id, axis=1)

In [171]:
full_merge_short

trip_id       stop_id  stop_sequence  \
0            7712.y1009.60-1-d12-1.1.O  8240DB000226              1   
1            7712.y1009.60-1-d12-1.1.O  8220DB000228              2   
2            7712.y1009.60-1-d12-1.1.O  8240DB000229              3   
3            7712.y1009.60-1-d12-1.1.O  8240DB000227              4   
4            7712.y1009.60-1-d12-1.1.O  8240DB000230              5   
...                                ...           ...            ...   
1874914  1468.y1003.60-84A-b12-1.342.I  8250DB003155             29   
1874915  1468.y1003.60-84A-b12-1.342.I  8250DB003156             30   
1874916  1468.y1003.60-84A-b12-1.342.I  8250DB003158             31   
1874917  1468.y1003.60-84A-b12-1.342.I  8250DB003159             32   
1874918  1468.y1003.60-84A-b12-1.342.I  8250DB003160             33   

             stop_headsign                        stop_name   stop_lat  \
0               Sandymount         Shanard Avenue, stop 226  53.391141   
1               Sandymount          Shanliss Road, stop 228  53.391877   
2               Sandymount           Oldtown Road, stop 229  53.391400   
3               Sandymount         Shanliss Drive, stop 227  53.391144   
4               Sandymount          Shanowen Road, stop 230  53.389888   
...                    ...                              ...        ...   
1874914   St Vincents Hosp       Beech Park Road, stop 3155  53.274256   
1874915   St Vincents Hosp            Deansworth, stop 3156  53.276538   
1874916   St Vincents Hosp      Deansgrange Road, stop 3158  53.278938   
1874917   St Vincents Hosp  Deansgrange Cemetery, stop 3159  53.282461   
1874918   St Vincents Hosp    St Fintan's Villas, stop 3160  53.284379   

         stop_lon  direction_id route_short_name  short_stop_id  
0       -6.262200             0                1            226  
1       -6.259720             0                1            228  
2       -6.256536             0                1            229  
3       -6.251345             0                1            227  
4       -6.249066             0                1            230  
...           ...           ...              ...            ...  
1874914 -6.161937             1              84a           3155  
1874915 -6.163024             1              84a           3156  
1874916 -6.164181             1              84a           3158  
1874917 -6.165920             1              84a           3159  
1874918 -6.166752             1              84a           3160  

[1874919 rows x 10 columns]

In [54]:
current_route_short = full_merge_short.route_short_name.tolist()

In [55]:
#converting all characters to uppercase to match the historical data 
x = map(lambda x:x.upper(),current_route_short)

In [56]:
current_route_short = list(x)

In [60]:
#Finding the new lines that have been added since 2018
diff_between_past_and_now = list(set(Past_routes) - set(current_route_short))
len(diff_between_past_and_now)

34

In [63]:
#Finding the new lines that have been added since 2018
diff_between_now_and_past = list(set(current_route_short) - set(Past_routes))
len(diff_between_now_and_past)

7

In [62]:
diff_between_past_and_now

['161',
 '45A',
 '42D',
 '33A',
 '114',
 '17A',
 '17',
 '270',
 '76A',
 '104',
 '75',
 '16C',
 '102',
 '14C',
 '63',
 '76',
 '31B',
 '59',
 '111',
 '32',
 '238',
 '185',
 '29A',
 '31D',
 '31A',
 '239',
 '220',
 '51X',
 '184',
 '70D',
 '31',
 '41A',
 '236',
 '18']

A number of these routes have been changed to Go-Ahead routes, so this GTFS data will ahve to be explored as well in order to find the stop sequnences. 

The list below shows the new routes that have been created 

29A -> H1  
32 -> H2  
31/ 31A/ 31B -> H3 and 6  

155, 66E and H9 are all new routes



In [64]:
diff_between_now_and_past

['H3', 'H1', '155', 'H2', '6', 'H9', '66E']

In [212]:
x = full_merge_short.loc[full_merge_short['route_short_name']== '16d']

In [213]:
x['trip_id'].iloc[0]

'1662.y1007.60-16D-d12-1.136.O'

In [66]:
route_list = full_merge_short.route_short_name.unique()

In [67]:
len(Current_routes)

143

In [70]:
len(route_list)

103

The missing routes here are accounted for in the Go-Ahead GTFS data and some are new routes which we will not be able to predict bus travel times for.

In [205]:
route_list

array(['1', '16', '16d', '33', '41', '41b', '41c', '41d', '44', '13',
       '11', '40', '40b', '40d', '122', '38', '38a', '38b', '38d', '53',
       '15a', '15b', '15d', '47', '56a', '61', '77a', '77x', '27', '120',
       '25', '25a', '25b', '25d', '26', '4', '66', '66a', '66b', '66e',
       '67', '7', '7a', '140', '155', '83', '83a', '9', '46a', '46e',
       '33e', '116', '37', '39', '39a', '70', '145', '33d', '7b', '7d',
       '33b', '32x', '41x', '84x', '142', '118', '25x', '39x', '66x',
       '67x', '27x', '44b', '14', '27b', '40e', '15', '150', '151', '33x',
       '51d', '68', '68a', '69', '69x', '123', '65', '65b', '68x', '49',
       '54a', '130', '27a', '6', 'H1', 'H2', 'H3', '42', '43', 'H9', '79',
       '79a', '84', '84a'], dtype=object)

In [243]:
##getting the stop sequence for each route in the up to date GTFS data 

stops_dict_current = {}

for route in route_list:
    print("//---------------------------------------------------//")
    print("route:", route)  
    sorted_df = full_merge_short.loc[full_merge_short['route_short_name']== '{}'.format(route)]

    if len(sorted_df.direction_id.unique()) == 1:
        
        direction = sorted_df.direction_id.unique()[0]
        
        one_direction = sorted_df.loc[sorted_df['direction_id'] == direction] 

        sorted_trip = one_direction.groupby(["trip_id"]).apply(lambda x: x.sort_values(["stop_sequence"], ascending = True)).reset_index(drop=True)

        sequence = sorted_trip.loc[sorted_trip['trip_id'] == sorted_trip.trip_id.iloc[0]].short_stop_id.tolist() 

        #converting all characters to uppercase to match the historical data 
        route = route.upper()

        stops_dict_current['{}_{}'.format(route,direction)] = {}
        stops_dict_current['{}_{}'.format(route,direction)] = sequence 

        print("{}_{} ".format(route, direction), "Complete")

    elif len(sorted_df.direction_id.unique()) == 2:
        
        for direction in range(0,2):   
            
            one_direction = sorted_df.loc[sorted_df['direction_id'] == direction]

            sorted_trip = one_direction.groupby(["trip_id"]).apply(lambda x: x.sort_values(["stop_sequence"], ascending = True)).reset_index(drop=True)
            trip = sorted_trip.trip_id.iloc[0]
            sequence = sorted_trip.loc[sorted_trip['trip_id'] == trip].short_stop_id.tolist() 

            #converting all characters to uppercase to match the historical data 
            route = route.upper()

            stops_dict_current['{}_{}'.format(route,direction)] = {}
            stops_dict_current['{}_{}'.format(route,direction)] = sequence 

            print("{}_{} ".format(route, direction), "Complete")


//---------------------------------------------------//
route: 1
1_0  Complete
1_1  Complete
//---------------------------------------------------//
route: 16
16_0  Complete
16_1  Complete
//---------------------------------------------------//
route: 16d
16D_0  Complete
//---------------------------------------------------//
route: 33
33_0  Complete
33_1  Complete
//---------------------------------------------------//
route: 41
41_0  Complete
41_1  Complete
//---------------------------------------------------//
route: 41b
41B_0  Complete
41B_1  Complete
//---------------------------------------------------//
route: 41c
41C_0  Complete
41C_1  Complete
//---------------------------------------------------//
route: 41d
41D_0  Complete
41D_1  Complete
//---------------------------------------------------//
route: 44
44_0  Complete
44_1  Complete
//---------------------------------------------------//
route: 13
13_0  Complete
13_1  Complete
//---------------------------------------------

123_0  Complete
123_1  Complete
//---------------------------------------------------//
route: 65
65_0  Complete
65_1  Complete
//---------------------------------------------------//
route: 65b
65B_0  Complete
65B_1  Complete
//---------------------------------------------------//
route: 68x
68X_1  Complete
//---------------------------------------------------//
route: 49
49_0  Complete
49_1  Complete
//---------------------------------------------------//
route: 54a
54A_0  Complete
54A_1  Complete
//---------------------------------------------------//
route: 130
130_0  Complete
130_1  Complete
//---------------------------------------------------//
route: 27a
27A_0  Complete
27A_1  Complete
//---------------------------------------------------//
route: 6
6_0  Complete
6_1  Complete
//---------------------------------------------------//
route: H1
H1_0  Complete
H1_1  Complete
//---------------------------------------------------//
route: H2
H2_0  Complete
H2_1  Complete
//----------

In [203]:
counter = 0 

one_direction = full_merge_short.loc[full_merge_short['route_short_name']== '33']
one_direction = one_direction.loc[one_direction['direction_id']== 1]
sorted_trip = one_direction.groupby(["trip_id"]).apply(lambda x: x.sort_values(["stop_sequence"], ascending = True)).reset_index(drop=True)

while len(sequence) != sequence_stops[-1:][0]:
    try:
        trip = sorted_trip.trip_id.iloc[counter]
        sequence = sorted_trip.loc[sorted_trip['trip_id'] == trip].short_stop_id.tolist() 
        sequence_stops  = sorted_trip.loc[sorted_trip['trip_id'] == trip].stop_sequence.tolist() 
        counter += len(sequence) 
        print(len(sequence))
        print(sequence_stops[-1:][0])
        
    except:
        trip = sorted_trip.trip_id.iloc[0]
        sequence = sorted_trip.loc[sorted_trip['trip_id'] == trip].short_stop_id.tolist()   
        print("wah")
        break
        

99
103
99
103
99
104
99
104
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
92
94
92
94
99
102
99
102
99
102
99
102
92
94
92
94
99
102
99
102
99
102
99
102
99
104
99
104
99
102
99
102
92
94
92
94
99
102
99
102
92
94
92
94
99
102
99
102
99
102
99
102
99
102
99
102
92
94
92
94
99
102
99
102
92
94
92
94
92
94
92
94
99
102
99
102
99
102
99
102
92
94
92
94
99
102
99
102
92
94
92
94
92
94
92
94
92
94
92
94
92
94
92
94
92
94
92
94
99
102
99
102
92
94
92
94
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
92
94
92
94
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
99
102
92
94
92
94
99
102
99
102
wah


we can see from the output above that the number of stops on each trip doesn't match the number of stops that are supposed to occur, this seems to be common across all of the routes.

In [5]:
with open('Stop_sequences_full_current_DUBLINBUS_GTFS.json', 'w', encoding='utf-8') as f:
    json.dump(stops_dict_current, f, ensure_ascii=False, indent=4)

NameError: name 'stops_dict_current' is not defined

In [245]:
f = open('Stop_sequences_full_current_GOAHEAD_GTFS.json',)
GTFS_GOAHEAD_JSON = json.load(f)

In [246]:
GTFS_GOAHEAD_JSON

{'161_0': [2934,
  2935,
  2936,
  2937,
  2938,
  7440,
  2940,
  2948,
  2949,
  2950,
  2954,
  7067,
  2956,
  2965,
  7448,
  2967,
  2968,
  2969,
  2970,
  4988,
  4707,
  4708,
  4336,
  4337,
  1310,
  1311,
  1312,
  1313,
  1055,
  2868,
  2825],
 '161_1': [2825,
  2867,
  1035,
  1314,
  1316,
  1315,
  4332,
  4333,
  4334,
  4712,
  4582,
  4987,
  2979,
  2980,
  2981,
  7447,
  2954,
  7067,
  2956,
  2965,
  2983,
  2984,
  7449,
  2927,
  2929,
  2930,
  7441,
  2933],
 '114_0': [6200,
  6201,
  6196,
  3532,
  3533,
  3491,
  6007,
  4349,
  3184,
  449,
  450,
  3181,
  5142,
  3187,
  3188,
  3189,
  3190,
  3191,
  3192,
  3193,
  3194,
  3195,
  3196,
  3198,
  3199,
  3200,
  3201,
  3085],
 '114_1': [3085,
  3082,
  3165,
  3166,
  3167,
  3168,
  3169,
  3171,
  3173,
  3174,
  3175,
  3176,
  3177,
  3178,
  3179,
  4984,
  448,
  4457,
  3182,
  3183,
  4350,
  5143,
  2833,
  6006,
  3493,
  3494,
  6197,
  6198,
  6199],
 '17_0': [1380,
  1381,
  1406,
  1

In [248]:
merged_JSON = {**stops_dict_current, **GTFS_GOAHEAD_JSON}

In [249]:
with open('Full_stops_sequence_merged.json', 'w', encoding='utf-8') as f:
    json.dump(merged_JSON, f, ensure_ascii=False, indent=4)